bigquery/params.go

```diff
@@ -127,14 +127,179 @@ type QueryParameter struct {
 	// above.  Null strings will report in query statistics as a valid empty
 	// string.
 	Value interface{}
+
+	explicitParameter Parameter
 }
 
-func (p QueryParameter) toBQ() (*bq.QueryParameter, error) {
+// ParameterType represents an explicit typed parameter to a query.
+// When using those interfaces, the priority is to avoid type inference.
+type ParameterType interface {
```

Comment:
It seems a bit odd to have an exported interface with no exported methods.

Historically we've tried to avoid leaking discovery types through as we don't control their implementation, and thus our guarantees about breaking changes etc.

Reply:
Yeah, I didn't like that too, but the ParameterType and ParameterValue interfaces both have methods that rely on the internal `bq.QueryParameterType` and `bq.QueryParameterValue`, which I was trying to avoid letting users know about it too much.